src/vm/util.cpp

```diff
@@ -1796,42 +1393,27 @@ size_t GetCacheSizePerLogicalCpu(BOOL bTrueSize)
     STATIC_CONTRACT_NOTHROW;
     STATIC_CONTRACT_GC_NOTRIGGER;
 
-    static size_t maxSize;
-    static size_t maxTrueSize;
+    static volatile size_t s_maxSize;
+    static volatile size_t s_maxTrueSize;
 
-    if (maxSize)
-    {
-        // maxSize and maxTrueSize cached
-        if (bTrueSize)
-        {
-            return maxTrueSize;
-        }
-        else
-        {
-            return maxSize;
-        }
-    }
+    size_t size = bTrueSize ? s_maxTrueSize : s_maxSize;
+    if (size != 0)
+        return size;
+
+    size_t maxSize, maxTrueSize;
 
-    // For x86, always get from cpuid.
-#if !defined (_TARGET_X86_)
     maxSize = maxTrueSize = GetLogicalProcessorCacheSizeFromOS() ; // Returns the size of the highest level processor cache
-#endif
 
-#if defined (_TARGET_X86) || defined(_TARGET_AMD64_)
```

Comment:
Note that the x86 ifdef is wrong - it missing trailing underscore. As result, `GetCacheSizePerLogicalCpu` always returns `0` on x86 today. This change makes it return cache size from OS. I hope that is fine with you. If you are worried about the behavior change, I can make it return hardcode `0`.

Reply:
ooh, thanks for noticing this! this definitely should not return 0. the fact it was should be fixed in 3.0. I will follow up.

Reply:
This was introduced recently by https://github.com/dotnet/coreclr/pull/24989/files#diff-8447e54277bb962d167a77bb260760d7R1820 .

Reply:
yes, I already talked to Andy about it.